src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -742,5 +748,103 @@ private void UpdateCacheExpirationTime(KeyValueWatcher changeWatcher, bool succe
 
             changeWatcher.CacheExpires = DateTimeOffset.UtcNow.Add(cacheExpirationTime);
         }
+
+        private async Task ExecuteWithFailOverPolicyAsync(Func<ConfigurationClient, Task> funcToExecute, CancellationToken cancellationToken = default)
+        {
+            using IEnumerator<ConfigurationClient> clientEnumerator = _configClientManager.GetAvailableClients().GetEnumerator();
+
+            // Guaranteed to return at least 1 element
+            Debug.Assert(clientEnumerator.MoveNext());
+
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    await funcToExecute(clientEnumerator.Current).ConfigureAwait(false);
+
+                    _configClientManager.UpdateClientStatus(clientEnumerator.Current, successful: true);
+
+                    break;
+                }
+                catch (AggregateException ae)
+                {
+                    if (!TryFailOver(ae, clientEnumerator))
+                    {
+                        throw;
+                    }
+                }
+                catch (RequestFailedException e)
+                {
+                    if (!TryFailOver(e, clientEnumerator))
+                    {
+                        throw;
+                    }
+                }
+            }
+        }
+
+        private bool TryFailOver(Exception e, IEnumerator<ConfigurationClient> clientEnumerator)
+        {
+            if (!ShouldFailoverForException(e))
+            {
+                return false;
+            }
+
+            _configClientManager.UpdateClientStatus(clientEnumerator.Current, successful: false);
+
+            if (!clientEnumerator.MoveNext())

```

Comment:
Splitting the MoveNext and the enumerator in two functions is error prone. I would call MoveNext in ExecuteWithFailOverPolicyAsync where the enumerator is created.

Reply:
I guess we will then end up with what you had before :). I was thinking about the catch with when clause to avoid duplicated code. For my education, what's the downside of that approach? We don't catch generic Exception but with restriction, right?

Reply:
I think the concern was that if we catch `Exception` and if there could be `OutOfMemoryException` etc then we would never realize it. But when we have restrictions, then this concern is not really a concern.

Reply:
If it's not really a concern, why don't we do it that way? The code will be simpler.

Reply:
If I had to choose in this case, I would rather have duplicated code but keep UpdateClientStatus and MoveNext in the same function.

Reply:
I added the `when` clause back. It makes the code less redundant and more readable. Also, since we have a `when` clause, we shouldn't be concerned about catching very broad exceptions.